AIVenture Project Blog Post

craigharman opened the pull request:
Hi Kevin,

Here is a first draft of the AIVenture blog article. I wanted to get your feedback before making any further changes. Please feel free to pick it apart as much as you like!

After re-reading it this morning I am wondering if a re-write to focus on the game engine and using that as the central point to then talk about user registration and tokens may be better?

I'm also conscious that I wasn't sure how technical or non-technical to go with this and so I think I've landed somewhere in the middle, not sure if that's where you want this to be?

Look forward to hearing from you.

phazonoverload commented on ai-in-directus-flows/index.md:
```diff
@@ -0,0 +1,154 @@
+
+Once the total game token limit is reached the prompt sent to Chat GPT is modified such that the next responses will include a satisfying conclusion to the game currently in progress.
+
+## The game engine
```
```suggestion
## The Game Engine
```

phazonoverload commented on ai-in-directus-flows/index.md:
```diff
@@ -0,0 +1,154 @@
+5. Providing a satisfactory end game state.
+6. State management for games in progress.
+
+In AIVenture all this logic is represented in Directus Flows. However with the large number of operations required we need to plan our flow architecture carefully to avoid unnecessarily complicating our game engine flow.
```
```suggestion
In AIVenture, all of this logic is represented in Directus Flows. However with the large number of operations required we need to plan our flow architecture carefully to avoid unnecessarily complicating our game engine flow.
```

phazonoverload commented on ai-in-directus-flows/index.md:
```diff
@@ -0,0 +1,154 @@
+
+In AIVenture all this logic is represented in Directus Flows. However with the large number of operations required we need to plan our flow architecture carefully to avoid unnecessarily complicating our game engine flow.
+
+### Managing complex flows.
```
```suggestion
### Managing Complex Flows
```

phazonoverload commented on ai-in-directus-flows/index.md:
```diff
@@ -0,0 +1,154 @@
+
+Creating a game engine in Directus Flows has the potential to become excessively large and unmaintainable very quickly without some strict architectural decisions. 
+
+We keep flows as small and functionally isolated components. Think of these flows as you would Classes in your application. We then have a central or main flow that triggers our sub-flows. This is a powerful and important architecture for our game engine design as it allows us to compartmentalise logic, keeping our main flow clean and readable and making testing easier. 
```
```suggestion
We keep flows as small and functionally-isolated components. Think of these flows as you would Classes in your application. A central (or 'main') flow triggers our sub-flows. This is a powerful and important architecture for our game engine design as it allows us to compartmentalize logic, keeping our main flow clean and readable and making testing easier. 
```

phazonoverload commented on ai-in-directus-flows/index.md:
```diff
@@ -0,0 +1,154 @@
+
+![Selecting “Another Flow” as our Trigger][image-1]
+
+Next we want to add our master or parent flow. This will be called from our frontend so should have a “Webhook” trigger - for ease of testing in the browser, let’s use the GET method. Then we add a new operation to our flow of type “Trigger Flow” and select our “SubCommand” flow from above. Our payload should include the data from the subcommand, ie. `{{ $last }}`.
```
```suggestion
Next, we want to add our central flow. This will be called from the frontend so should have a “Webhook” trigger. For ease of testing in the browser, use the GET method. Then, we add a new operation to our flow of type “Trigger Flow” and select our “SubCommand” flow from above. Our payload should include the data from the subcommand, ie. `{{ $last }}`.
```

Standardizing earlier language, but also `master` is a term we generally keep our of projects and writing where possible.

craigharman replied:
Understood. Perhaps "main" instead of "central" as this flow is really the top most piece of logic. Thoughts?

phazonoverload replied:
Yup works for me. It was more the rejection of the word master than a promotion of central. Whatever works for you.

phazonoverload commented on ai-in-directus-flows/index.md:
````diff
@@ -0,0 +1,154 @@
+}
+```
+
+In AIVenture we use this sub-flow technique extensively wherever we feel we have a self-contained functional component such as interacting with Chat GPT API, calculating token usage or processing user input.
````
```suggestion
In AIVenture, we use this sub-flow technique extensively wherever we feel we have a self-contained functional component such as interacting with Chat GPT API, calculating token usage or processing user input.
```

phazonoverload commented on ai-in-directus-flows/index.md:
```diff
@@ -0,0 +1,154 @@
+
+In AIVenture we use this sub-flow technique extensively wherever we feel we have a self-contained functional component such as interacting with Chat GPT API, calculating token usage or processing user input.
+
+### No turning back
```
```suggestion
### Branching Logic With Only Two Paths
```

phazonoverload commented on ai-in-directus-flows/index.md:
```diff
@@ -0,0 +1,154 @@
+
+The underlying application is effectively a Chat-GPT powered game engine. It uses Directus to control game logic, provide player registration, and track game state persistence. The front-end is a [Nuxt][5] application which receives player commands (in the form of text prompts) and sends them to a Directus Flow endpoint. Responses are generated by Chat-GPT using the [directus-extension-ai-pack][6 ...
+
+## On boarding players
```
```suggestion
## Registering Players
```

phazonoverload commented on ai-in-directus-flows/index.md:
```diff
@@ -0,0 +1,154 @@
+The underlying application is effectively a Chat-GPT powered game engine. It uses Directus to control game logic, provide player registration, and track game state persistence. The front-end is a [Nuxt][5] application which receives player commands (in the form of text prompts) and sends them to a Directus Flow endpoint. Responses are generated by Chat-GPT using the [directus-extension-ai-pack][6 ...
+
+## On boarding players
+So that AIVenture can store game progress across devices individual users need to be identified. 
```
```suggestion
So that AIVenture can store game progress across devices, individual users need to be identified. 
```

phazonoverload commented on ai-in-directus-flows/index.md:
```diff
@@ -0,0 +1,154 @@
+## On boarding players
+So that AIVenture can store game progress across devices individual users need to be identified. 
+
+As Directus already has user management and authentication baked in AIVenture is able to harness the API to create users. When users sign up they are assigned a custom “player” role that gives them permissions to reach a single Directus Flow game endpoint.
```
```suggestion
As Directus already has user management and authentication baked in, AIVenture is able to harness the API to create users. When users sign up, they are assigned a custom “player” role that gives them permissions to reach a single Directus Flow game endpoint.
```

phazonoverload commented on ai-in-directus-flows/index.md:
```diff
@@ -0,0 +1,154 @@
+
+As Directus already has user management and authentication baked in AIVenture is able to harness the API to create users. When users sign up they are assigned a custom “player” role that gives them permissions to reach a single Directus Flow game endpoint.
+
+To avoid the game sign-up process being spammed with fake email addresses we require that a player’s email address is validated. Although there isn’t specifically an email validation workflow available in Directus, there is a [user invite procedure][7]. Behind the scenes when a player signs up to AIVenture we are actually generating a user **invite** in Directus. This does not give access to the  ...
```
```suggestion
To avoid the game sign-up process being spammed with fake email addresses, we require that a player’s email address is validated. Although there isn’t specifically an email validation workflow available in Directus, there is a [user invite procedure][7]. Behind the scenes when a player signs up to AIVenture we are actually generating a user **invite** in Directus. This does not give access to the game until it has been accepted via a link in the player’s email - which is effectively the same as an email validation process.
```

phazonoverload commented on ai-in-directus-flows/index.md:
```diff
@@ -0,0 +1,154 @@
+
+To avoid the game sign-up process being spammed with fake email addresses we require that a player’s email address is validated. Although there isn’t specifically an email validation workflow available in Directus, there is a [user invite procedure][7]. Behind the scenes when a player signs up to AIVenture we are actually generating a user **invite** in Directus. This does not give access to the  ...
+
+The emails sent to players are customised by directly editing the [liquid][8] template [files][9] provided by Directus. Only upon successfully completing the invite procedure does a user get provided with an active Directus user account and can hence access the game.
```
```suggestion
The emails sent to players are customised by directly editing the [liquid][8] template [files][9] provided by Directus. Once a user accepts their invite, they are provided with an active Directus user account and can access the game.
```

phazonoverload commented on ai-in-directus-flows/index.md:
```diff
@@ -0,0 +1,154 @@
+## Tracking OpenAI Tokens
+AIVenture uses the [directus-extension-ai-pack][10] extension to interact with Open AI’s **Chat Completion API** within Directus Flows. The extension requires minimal setup before allowing chat prompts to be sent to Open AI and have responses returned for further processing in our flows.
+
+Chat GPT charges the API account owner for usage via [tokens][11] allocated by the length of your prompts and the resulting responses. In order to keep Open AI bills under control AIVenture allocates daily, per user, token limits on a player’s progress through the game as well as a hard limit that indicates the game needs to reach a conclusion.
```
```suggestion
OpenAI charges the API account owner for usage via [tokens][11] allocated by the length of your prompts and the resulting responses. In order to keep OpenAI bills under control, AIVenture allocates daily, per user, token limits on a player’s progress through the game as well as a hard limit that indicates the game needs to reach a conclusion.
```

phazonoverload commented on ai-in-directus-flows/index.md:
```diff
@@ -0,0 +1,154 @@
+The emails sent to players are customised by directly editing the [liquid][8] template [files][9] provided by Directus. Only upon successfully completing the invite procedure does a user get provided with an active Directus user account and can hence access the game.
+
+## Tracking OpenAI Tokens
+AIVenture uses the [directus-extension-ai-pack][10] extension to interact with Open AI’s **Chat Completion API** within Directus Flows. The extension requires minimal setup before allowing chat prompts to be sent to Open AI and have responses returned for further processing in our flows.
```
```suggestion
AIVenture uses the [directus-extension-ai-pack][10] extension to interact with OpenAI’s **Chat Completion API** within Directus Flows. The extension requires minimal setup before allowing chat prompts to be sent to OpenAI and have responses returned for further processing in our flows.
```

phazonoverload commented on ai-in-directus-flows/index.md:
```diff
@@ -0,0 +1,154 @@
+
+Chat GPT charges the API account owner for usage via [tokens][11] allocated by the length of your prompts and the resulting responses. In order to keep Open AI bills under control AIVenture allocates daily, per user, token limits on a player’s progress through the game as well as a hard limit that indicates the game needs to reach a conclusion.
+
+The Chat GPT API returns a token count with each response, however this was originally not passed on by [directus-extension-ai-pack][12]. Luckily, the extension is open-source and [a small PR later][13] we were able to access the Chat GPT usage data in our flow:
```
```suggestion
The ChatGPT API returns a token count with each response, however this was originally not passed on by [directus-extension-ai-pack][12]. Luckily, the extension is open-source and [a small PR later][13] we were able to access the Chat GPT usage data in our flow:
```

phazonoverload commented on ai-in-directus-flows/index.md:
```diff
@@ -0,0 +1,154 @@
+
+Each prompt/response combination is stored in a Directus collection and various query filters and aggregators are used to provide the game engine with the total number of tokens used by a player in a day and the total tokens used for the current game.
+
+If the player reaches their daily limit, no more Chat GPT prompts can be sent to the API and the game will prompt you to wait until tomorrow before being able to progress.
```
```suggestion
If the player reaches their daily limit, no more prompts can be sent to the API and the game will prompt them to wait until tomorrow before being able to progress.
```

phazonoverload commented on ai-in-directus-flows/index.md:
```diff
@@ -0,0 +1,154 @@
+
+If the player reaches their daily limit, no more Chat GPT prompts can be sent to the API and the game will prompt you to wait until tomorrow before being able to progress.
+
+Once the total game token limit is reached the prompt sent to Chat GPT is modified such that the next responses will include a satisfying conclusion to the game currently in progress.
```
```suggestion
Once the total game token limit is reached, the prompt sent to ChatGPT is modified such that the next responses will include a satisfying conclusion to the game currently in progress.
```

phazonoverload reviewed (commented):
This is really, really good and a very enjoyable read. Completely agree with your assessment that the post should start with the game engine and them move into registration and tracking token usage. 

Other than that - a few simplifications in language, adding punctuation, and fixing brand names. Lots of changes, but they are tiny.

craigharman commented:
Thanks for the positive feedback @phazonoverload !

I've added all the requested changes, added the screenshot and relocated the game engine to the top which I think works a lot better. Back to you!

phazonoverload commented:
Much happier with this and happy to accept it as-is. I can apply the change to `main` in that suggestion when publishing, along with standardizing to 'ChatGPT'. 

Only other thing I need is a third-person bio for yourself. For example, mine is: 

> Kevin runs Developer Relations at Directus and is Director of You Got This - a learning hub to help developers improve their core skills. He is an avid boardgamer, tired dad, and shameless Disney adult. He is based in Berlin with his partner and two kids. 

From here, I will get artwork made and schedule this in - probably for publishing on October 3. 

This repo is just for authoring/reviewing - the actual piece will be published via a Directus project, so I'll take it from here. I'll close this issue to keep the main branch clean, but this is a acceptance of your work. 

Could you email me at kevin@directus.io and I'll share what we need to get you paid. 

I'm very excited to publish this piece - it's very cool.
